### ISSUE TITLE
WebSocket timeouts 

### ISSUE THREAD

**Issue by gyalai** [rancher/rancher#8732]
**Rancher Versions:**
Server: 1.6.0
healthcheck: 0.3.0
ipsec: 0.1.0
network-services: 0.2.0

**Docker Version:**
The server running on the 17.05.0-ce version and the hosts are running 17.05.0-ce version or 17.03.0-ce

**OS and where are the hosts located? (cloud, bare metal, etc):**
Bare Matel Ubuntu 16.04 hosts with the 4.4.0-77 kernel.

An nginx proxy run infront of the rancher server.

After the server has been updated to the 1.6.0 version from the 1.5.1 version. We have lots of log message like this 
`2017/05/10 15:52:55 [error] 6931#6931: *4055131 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.10.3.5, server: rancher.server.no, request: "GET /v1/subscribe?eventNames=ping&eventNames=config.update HTTP/1.1", upstream: "http://127.0.0.1:5000/v1/subscribe?eventNames=ping&eventNames=config.update", host: "rancher.server.no"
2017/05/10 15:52:55 [error] 6931#6931: *4055137 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.10.3.6, server: rancher.server.no, request: "GET /v1/subscribe?eventNames=ping&eventNames=config.update HTTP/1.1", upstream: "http://127.0.0.1:5000/v1/subscribe?eventNames=ping&eventNames=config.update", host: "rancher.server.no"
2017/05/10 15:52:55 [error] 6923#6923: *4055140 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.10.2.23, server: rancher.server.no, request: "GET /v1/subscribe?eventNames=ping&eventNames=config.update HTTP/1.1", upstream: "http://127.0.0.1:5000/v1/subscribe?eventNames=ping&eventNames=config.update", host: "rancher.server.no"
2017/05/10 15:52:58 [error] 6913#6913: *4055320 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.8.0.10, server: rancher.server.no, request: "GET /v2-beta/projects/1a10/subscribe?eventNames=resource.change&limit=-1&sockId=53 HTTP/1.1", upstream: "http://127.0.0.1:5000/v2-beta/projects/1a10/subscribe?eventNames=resource.change&limit=-1&sockId=53", host: "rancher.server.no"
2017/05/10 15:53:01 [error] 6932#6932: *4055280 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.8.0.18, server: rancher.server.no, request: "GET /v2-beta/projects/1a11/subscribe?eventNames=resource.change&limit=-1&sockId=40 HTTP/1.1", upstream: "http://127.0.0.1:5000/v2-beta/projects/1a11/subscribe?eventNames=resource.change&limit=-1&sockId=40", host: "rancher.server.no"
2017/05/10 15:53:02 [error] 6932#6932: *4055286 recv() failed (104: Connection reset by peer) while proxying upgraded connection, client: 10.8.0.10, server: rancher.server.no, request: "GET /v2-beta/projects/1a32/subscribe?eventNames=resource.change&limit=-1&sockId=46 HTTP/1.1", upstream: "http://127.0.0.1:5000/v2-beta/projects/1a32/subscribe?eventNames=resource.change&limit=-1&sockId=46", host: "rancher.server.no"`

In the rancher-server log:
`2017-05-10 13:52:49,509 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [1] 
2017-05-10 13:52:49,514 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [4] count [6] 
2017-05-10 13:52:49,514 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [4] 
2017-05-10 13:52:49,517 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Failed to get ping from agent [692] count [6] 
2017-05-10 13:52:49,517 ERROR [:] [] [] [] [TaskScheduler-2] [i.c.p.a.s.ping.impl.PingMonitorImpl ] Scheduling reconnect for [692] 
time="2017-05-10T13:52:50Z" level=info msg="Shutting down backend 2c9dbbef-5b47-4426-ab0c-1eb5ce23d506. Connection closed because: read tcp 172.17.0.2:8080->172.17.0.1:50902: read: connection reset by peer." 
time="2017-05-10T13:52:50Z" level=info msg="Removed backend. Key: 2c9dbbef-5b47-4426-ab0c-1eb5ce23d506. Session ID 3a9b1251-2cfe-4d56-b2ef-b69cadeacc0d ." 
time="2017-05-10T13:52:50Z" level=info msg="Shutting down backend e57a4631-6da6-49ca-99c9-d0dc6add95ec. Connection closed because: read tcp 172.17.0.2:8080->172.17.0.1:50654: read: connection reset by peer." 
time="2017-05-10T13:52:50Z" level=info msg="Removed backend. Key: e57a4631-6da6-49ca-99c9-d0dc6add95ec. Session ID 70a3089c-0038-4754-a520-2f6b5642111a ." 
time="2017-05-10T13:52:52Z" level=info msg="Shutting down backend 9d1b7c9f-a395-4331-5427-c56e67efe850. Connection closed because: read tcp 172.17.0.2:8080->172.17.0.1:50438: read: connection reset by peer." 
time="2017-05-10T13:52:52Z" level=info msg="Removed backend. Key: 9d1b7c9f-a395-4331-5427-c56e67efe850. Session ID 11ca220a-a396-4f9b-8517-7228feb7be94 ." 
time="2017-05-10T13:52:53Z" level=info msg="Shutting down backend 3719e2a5-01dc-44c0-9a6a-9b9b096764b1. Connection closed because: read tcp 172.17.0.2:8080->172.17.0.1:50500: read: connection reset by peer." 
time="2017-05-10T13:52:53Z" level=info msg="Removed backend. Key: 3719e2a5-01dc-44c0-9a6a-9b9b096764b1. Session ID cd49cb2f-881f-43e7-ba76-a85b2dc196dc ." 
time="2017-05-10T13:52:54Z" level=info msg="Shutting down backend 9a1b8526-fd8c-4909-50de-fed491ac14cb. Connection closed because: websocket: close 1006 unexpected EOF."`

Because these websocket connection loss: 
1. The hosts in the environment have became Disconnected and we have to restart the rancher-agent on the host. 
2. During a stack update we sometimes got "Time out getting ip" issue and then we have to restart the network-services/metadata service. After the restart the stack get updated properly.

Any help would be appreciated

**Comment by cjellick** [comment]
One typical cause of this is that either the proxy (in your case nginx) is not configured properly to handle websocket connections. Could you share your nginx config (sanitized to remove sensitivie data)

**Comment by gyalai** [comment]
Yes, that was my initial thought too but we are using the same configuration for months. 


We are running the rancher-server on the same host under the 5000 port.

This is the configuration what we are using.

```
 upstream rancher {
    server 127.0.0.1:5000;
}

server {
    listen 443 ssl spydy;
    server_name rancher.server.no;

    # set HSTS-Header because we only allow https traffic
    add_header Strict-Transport-Security "max-age=31536000; includeSubdomains";

    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_pass http://rancher;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        # This allows the ability for the execute shell window to remain open for up to 15 minutes. Without this parameter, the default is 1 minute and will automatically close.
        proxy_read_timeout 900s;
    }
}

server {
    listen 80;
    server_name rancher.server.no;
    return 301 https://$server_name$request_uri;
}
```

Maybe, this part could be interesting too.

```
user www-data;
worker_processes auto;
pid /run/nginx.pid;

events {
	worker_connections 768;
	# multi_accept on;
}
```

**Comment by gyalai** [comment]
Hi,

I think I solved the issue. 

I made some tcpdump and looked for the rst package. I wanted to figure out is the client, the nginx or the server send that first and initiate the websocket close. It turned out none of them. The rst package was sent from the docker gateway ip first to the rancher server.

So, my solution was to create a separate bridged network for the rancher server. It run without any websocket issue for 1 day.
